File: TcpSocketSystem_DotNet/src/Infrastructure/Network/TcpServidorSocket.cs
using System.Net;
using System.Net.Sockets;
using TcpSocketSystem.Core.Domain.Puertos;

namespace TcpSocketSystem.Core.Infrastructure.Network;

/// <summary>
/// Servidor TCP asíncrono y multihilo basado en TcpListener y Task.Run.
/// Gestiona concurrencia, captura puertos ocupados y previene bloqueos de cola.
/// </summary>
public sealed class TcpServidorSocket : IServicioRed
{
    private readonly IServicioLogging _logger;
    private TcpListener? _listener;
    private bool _estaEscuchando;
    private int _puertoAsignado;

    public bool EstaEscuchando => _estaEscuchando;
    public int PuertoAsignado => _puertoAsignado;

    public TcpServidorSocket(IServicioLogging logger)
    {
        _logger = logger ?? throw new ArgumentNullException(nameof(logger));
    }

    public async Task IniciarServidorAsync(
        int puerto,
        Func<ISocketConexion, CancellationToken, Task> onClienteConectado,
        CancellationToken cancellationToken)
    {
        if (_estaEscuchando)
        {
            throw new InvalidOperationException("El servidor de sockets ya se encuentra en ejecución.");
        }

        try
        {
            _puertoAsignado = puerto;
            _listener = new TcpListener(IPAddress.Any, puerto);
            _listener.Start();
            _estaEscuchando = true;

            _logger.Info($"Servidor TCP iniciado exitosamente y escuchando en 0.0.0.0:{puerto}", nameof(TcpServidorSocket));
            _logger.Auditoria("SERVIDOR_INICIADO", $"0.0.0.0:{puerto}", "Servidor listo para aceptar conexiones entrantes.");

            while (!cancellationToken.IsCancellationRequested)
            {
                try
                {
                    // Aceptación asíncrona no bloqueante
                    var tcpClient = await _listener.AcceptTcpClientAsync(cancellationToken);
                    
                    // Configuración de sockets para optimizar rendimiento y resiliencia
                    tcpClient.NoDelay = true; // Deshabilita algoritmo de Nagle para baja latencia
                    tcpClient.ReceiveTimeout = 30000; // 30 segundos timeout
                    tcpClient.SendTimeout = 10000;

                    var conexion = new TcpConexionWrapper(tcpClient);

                    // Despachar en hilo de trabajo independiente (Multicliente sin bloqueo)
                    _ = Task.Run(async () =>
                    {
                        try
                        {
                            await onClienteConectado(conexion, cancellationToken);
                        }
                        catch (Exception ex)
                        {
                            _logger.Error($"Error no controlado en la atención del cliente {conexion.DireccionRemota}:{conexion.PuertoRemoto}", ex, nameof(TcpServidorSocket));
                        }
                        finally
                        {
                            await conexion.DisposeAsync();
                        }
                    }, cancellationToken);
                }
                catch (OperationCanceledException)
                {
                    break;
                }
                catch (SocketException) when (!_estaEscuchando)
                {
                    // El listener se detuvo deliberadamente
                    break;
                }
                catch (Exception ex)
                {
                    _logger.Error("Error al aceptar una nueva conexión entrante.", ex, nameof(TcpServidorSocket));
                }
            }
        }
        catch (SocketException ex) when (ex.SocketErrorCode == SocketError.AddressAlreadyInUse)
        {
            _logger.Error($"No se pudo iniciar el servidor. El puerto {puerto} ya se encuentra ocupado por otro proceso.", ex, nameof(TcpServidorSocket));
            throw;
        }
        catch (Exception ex)
        {
            _logger.Error($"Fallo crítico al iniciar el servidor en el puerto {puerto}.", ex, nameof(TcpServidorSocket));
            throw;
        }
        finally
        {
            await DetenerServidorAsync();
        }
    }

    public Task DetenerServidorAsync()
    {
        if (!_estaEscuchando) return Task.CompletedTask;

        _estaEscuchando = false;
        try
        {
            _listener?.Stop();
            _logger.Info("Servidor TCP detenido de forma ordenada.", nameof(TcpServidorSocket));
            _logger.Auditoria("SERVIDOR_DETENIDO", $"0.0.0.0:{_puertoAsignado}", "Servidor finalizó la escucha.");
        }
        catch (Exception ex)
        {
            _logger.Error("Error al detener el listener de sockets.", ex, nameof(TcpServidorSocket));
        }

        return Task.CompletedTask;
    }
}

File: TcpSocketSystem_Java/src/com/tecnm/sockets/infrastructure/network/TcpServidorSocket.java
package com.tecnm.sockets.infrastructure.network;

import com.tecnm.sockets.domain.puertos.IServicioLogging;
import com.tecnm.sockets.domain.puertos.IServicioRed;
import com.tecnm.sockets.domain.puertos.ISocketConexion;
import java.io.IOException;
import java.net.BindException;
import java.net.InetSocketAddress;
import java.net.ServerSocket;
import java.net.Socket;
import java.util.concurrent.atomic.AtomicBoolean;
import java.util.function.Consumer;

/**
 * Servidor TCP multihilo en Java 21 utilizando Virtual Threads (Project Loom)
 * para atender concurrencia masiva de sockets de forma ultra-ligera sin bloqueo de hilos de SO.
 */
public final class TcpServidorSocket implements IServicioRed {

    private final IServicioLogging logger;
    private ServerSocket serverSocket;
    private final AtomicBoolean estaEscuchando = new AtomicBoolean(false);
    private int puertoAsignado;

    public TcpServidorSocket(IServicioLogging logger) {
        if (logger == null) {
            throw new IllegalArgumentException("El servicio de logging no puede ser nulo.");
        }
        this.logger = logger;
    }

    @Override
    public boolean estaEscuchando() {
        return estaEscuchando.get();
    }

    @Override
    public int getPuertoAsignado() {
        return puertoAsignado;
    }

    @Override
    public void iniciarServidor(int puerto, Consumer<ISocketConexion> onClienteConectado) throws Exception {
        if (estaEscuchando.get()) {
            throw new IllegalStateException("El servidor ya está en ejecución.");
        }

        try {
            this.puertoAsignado = puerto;
            this.serverSocket = new ServerSocket();
            this.serverSocket.setReuseAddress(true);
            this.serverSocket.bind(new InetSocketAddress(puerto));
            this.estaEscuchando.set(true);

            logger.info(String.format("Servidor TCP Java (Virtual Threads) iniciado en 0.0.0.0:%d", puerto), "TcpServidorSocket");
            logger.auditoria("SERVIDOR_INICIADO", "0.0.0.0:" + puerto, "Listo para atender conexiones.");

            while (estaEscuchando.get() && !serverSocket.isClosed()) {
                try {
                    Socket socket = serverSocket.accept();
                    socket.setTcpNoDelay(true);
                    socket.setSoTimeout(30000); // 30 segundos timeout

                    TcpSocketConexion conexion = new TcpSocketConexion(socket);

                    // Concurrencia de alta eficiencia (Virtual Threads en Java 21 / Fallback multiplataforma)
                    GestorHilos.iniciarHilo("vt-cliente-" + conexion.getId(), () -> {
                        try (TcpSocketConexion conn = conexion) {
                            onClienteConectado.accept(conn);
                        } catch (Exception ex) {
                            logger.error(String.format("Error en la atención del cliente %s:%d",
                                    conexion.getDireccionRemota(), conexion.getPuertoRemoto()), ex, "TcpServidorSocket");
                        }
                    });

                } catch (IOException ex) {
                    if (!estaEscuchando.get() || serverSocket.isClosed()) {
                        break;
                    }
                    logger.error("Error al aceptar conexión entrante.", ex, "TcpServidorSocket");
                }
            }
        } catch (BindException ex) {
            logger.error(String.format("No se pudo iniciar el servidor. El puerto %d ya está en uso.", puerto), ex, "TcpServidorSocket");
            throw ex;
        } finally {
            detenerServidor();
        }
    }

    @Override
    public void detenerServidor() {
        if (!estaEscuchando.getAndSet(false)) {
            return;
        }

        try {
            if (serverSocket != null && !serverSocket.isClosed()) {
                serverSocket.close();
                logger.info("Servidor TCP Java detenido limpiamente.", "TcpServidorSocket");
                logger.auditoria("SERVIDOR_DETENIDO", "0.0.0.0:" + puertoAsignado, "Servidor cerró el listener.");
            }
        } catch (IOException ex) {
            logger.error("Error al cerrar el ServerSocket.", ex, "TcpServidorSocket");
        }
    }
}
